### PR TITLE
Handle Task/Monitor Change Events

### PR DESCRIPTION
**Pull request by jjbuchan**
# What

Adds consumers for `TaskChangeEvent`s and `MonitorChangeEvent`s.

* `TaskChangeEvent`s cause (re/un)deploys of task queries to esper
* `MonitorChangeEvent`s cause monitor property caches to be cleared

# How

1. Updated `CachedRepositoryRequests.java` to have some new `CacheEvict` methods
1. The "normal" kafka consumer properties in `application.yml` are used for these new consumers
    * The metrics consumer has been updated to provide its own custom properties
1. The partition tracking logic for the metrics consumer has been moved to its own class `PartitionTracker.java`
   * This enables multiple classes to view what partitions the consumer is responsible for
1. Moved some test logic to `MultipleConsumerTestConfig.java` since it started getting cluttered due to needing 3 instances of multiple classes

## How to test

Run all tests.

I will begin manual testing tomorrow to verify that all recent pieces of event-processor / mgmt are working and to try and identify any missing parts.

# Why

Tasks are currently only loaded at startup.  This enables it to keep track of new changes after it is already running.

Caching of monitor data was added to limit the number of db lookups.  This now clears those caches if the monitor has been modified.

### CODE REVIEW

File: src/main/java/com/rackspace/salus/event/processor/services/ChangeEventsListener.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    Pageable pageRequest = PageRequest.of(0, 1000);
+    Page<BoundMonitor> boundMonitors = boundMonitorRepository
+        .findAllByMonitor_IdAndMonitor_TenantId(event.getMonitorId(), event.getTenantId(), pageRequest);
+
+    while (boundMonitors.hasContent()) {
+      // try to limit cache evict calls by filtering the duplicate keys.
+      // each of these will already contain the same tenantId and monitorId so we only need to
+      // perform a distinct operation on the resourceId.
+      Collection<BoundMonitor> uniqueKeys = boundMonitors.get()
+          .collect(Collectors.groupingBy(BoundMonitor::getResourceId,
```
Comment (GeorgeJahad) [inline src/main/java/com/rackspace/salus/event/processor/services/ChangeEventsListener.java:100]: This code is a bit hard to follow.  It is first creating a map of resourceIds to lists of boundMonitors, right?

Then it converts that to a list of bound monitors one for each resource?

Comment (jjbuchan) [reply src/main/java/com/rackspace/salus/event/processor/services/ChangeEventsListener.java:100]: That's correct.  The end goal is that I only want to have one item per `tenantId:resourceId:monitorId`.  Due to the paginated calls there may be some overlap but doing a "best effort" seemed worthwhile.

File: src/main/java/com/rackspace/salus/event/processor/services/PartitionTracker.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    Set<Integer> removedPartitions = new HashSet<>(trackedPartitions);
+    removedPartitions.removeAll(assignedPartitions);
+
+    Set<Integer> addedPartitions = new HashSet<>(assignedPartitions);
+    addedPartitions.removeAll(trackedPartitions);
+
+    trackedPartitions = assignedPartitions;
+
+    handler.removeTasksForPartitions(removedPartitions);
+    handler.deployTasksForPartitions(addedPartitions);
```
Comment (GeorgeJahad) [inline src/main/java/com/rackspace/salus/event/processor/services/PartitionTracker.java:92]: haven't tried to think through this yet, but i'm wondering if you have:

 What is the likelihood of race conditions on partition reload, (especially at startup,)  with change events happening at the same time.  Should change event handling be paused during reloads?

Comment (jjbuchan) [reply src/main/java/com/rackspace/salus/event/processor/services/PartitionTracker.java:92]: That's a good question.  With old-esper that probably wouldn't be a problem since the query names are unique, but here we probably do risk adding duplicates.

I'll look into it more and likely add that logic in a follow up PR.

We'd probably want to have the consumer start reading change event topics at the moment after sql is initially read upon start, and not wait until when it has completed loading everything.  Maybe there will have to be additional logic in multiple places to ensure we don't miss anything.

I guess that's an extra benefit of duplicating regions in the past - if one region missed it the other two should still get it.

**Review by GeorgeJahad — APPROVED** [review]
